archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,302 @@
+      simp only [not_lt] at p,
+      interval_cases l; linarith }},
+  -- Otherwise, if 100 ≤ n < 107, then it suffices to consider explicit
+  -- construction of a triplet {a, b, c}, which is constructed by setting l=9
+  -- in the argument at the start of the file.
+  { refine ⟨126, 163, 198, _, _, _, _,⟨17, _⟩, ⟨18, _⟩, 19, _⟩; linarith },
+end
+
+-- Since it will be more convenient to work with sets later on, we will translate the above claim
+-- to state that there always exists a set B ⊆ [n, 2n] of cardinality at least 3, such that each
+-- pair of pairwise unequal elements of B sums to a perfect square.
+lemma exists_finset_of_3_leq_card_with_pairs_summing_to_squares (n : ℕ) (hn : 100 ≤ n) :
+  ∃ (B : finset ℕ), 2 * 1 + 1 ≤ B.card ∧
+  (∀ (a b ∈ B), a ≠ b → ∃ k, a + b = k * k) ∧
+  ∀ (c ∈ B), n ≤ c ∧ c ≤ 2 * n :=
+begin
+  obtain ⟨a, b, c, hna, hab, hbc, hcn, h₁, h₂, h₃⟩ := exists_triplet_summing_to_squares n hn,
+  refine ⟨{a, b, c}, _, _, _⟩,
+  { suffices : ({a, b, c} : finset ℕ).card = 3,
+    { linarith },
+    rw [finset.card_insert_of_not_mem, finset.card_insert_of_not_mem, finset.card_singleton],
+    { rw finset.mem_singleton,
+      exact hbc.ne },
+    { simp only [finset.mem_insert, finset.mem_singleton],
+      push_neg,
+      exact ⟨hab.ne, (hab.trans hbc).ne⟩ }},
+  { intros x y hx hy hxy,
+    simp only [finset.mem_insert, finset.mem_singleton] at hx,
+    simp only [finset.mem_insert, finset.mem_singleton] at hy,
+    rcases hx with (hxa | hxb | hxc),
+    { rcases hy with (hya | hyb | hyc),
+      { rw ← hya at hxa,
+        contradiction },
+      { convert h₁,
+        rw [hxa, hyb] },
+      { convert h₂,
+        rw [hxa, hyc, add_comm] }},
+    { rcases hy with (hya | hyb | hyc),
+      { convert h₁,
+        rw [hxb, hya, add_comm] },
+      { rw ← hyb at hxb,
+        contradiction },
+      { convert h₃,
+        rw [hxb, hyc] }},
+    { rcases hy with (hya | hyb | hyc),
+      { convert h₂,
+        rw [hxc, hya] },
+      { convert h₃,
+        rw [hxc, hyb, add_comm] },
+      { rw ← hyc at hxc,
+        contradiction }}},
+  { intros d hd,
+    simp only [finset.mem_insert, finset.mem_singleton] at hd,
+    split,
+    { rcases hd with (hd | hd | hd);
+      { linarith }},
+    { rcases hd with (hd | hd | hd);
+      { linarith }}},
+end
+
+-- If for n ∈ ℕ, and finite sets B, X, Y such that B ⊆ X ∪ Y and B has cardinality at least
+-- 2 * n + 1, then there exists a subset C ⊆ A of cardinality at least n + 1, which is entirely
+-- contained in X or Y. For our result, we will apply this lemma with n = 1 and any X such that
+-- X ⊆ [n, 2n] and Y = [n, 2n] \ X, while we'll let B ⊆ [n, 2n] be the subset such that each
+-- pairwise unequal pair of B sums to a perfect square.
+lemma finset_subset_or_complement_cardinality {α : Type*} [decidable_eq α] (B X Y : finset α)
+  (h : B ⊆ X ∪ Y) {n : ℕ} (hB : 2 * n + 1 ≤ B.card) :
+  (∃ C, C ⊆ B ∧ C ⊆ X ∧ n + 1 ≤ C.card) ∨ (∃ C, C ⊆ B ∧ C ⊆ Y ∧ n + 1 ≤ C.card) :=
```

Comment:
This is still too specialized to be proving directly without making use of an existing statement of the pigeonhole principle.

Reply:
Understood! Should I try to come up with a proof using the pigeonhole principle, or should I use an entirely different approach instead?

Reply:
Keeping the statement basically the same, you could fill in the sorries in
```
lemma snorg {a b n : ℕ} (h : 2 * n + 1 ≤ a + b) : n + 1 ≤ a ∨ n + 1 ≤ b :=
begin
  by_contradiction w,
  simp [not_or_distrib] at w,
  cases w with w₁ w₂,
  linarith,
end

lemma finset_subset_or_complement_cardinality {α : Type*} [decidable_eq α] (B X Y : finset α)
  (h : B ⊆ X ∪ Y) {n : ℕ} (hB : 2 * n + 1 ≤ B.card) :
  ∃ C, C ⊆ B ∧ n + 1 ≤ C.card ∧ (C ⊆ X ∨ C ⊆ Y) :=
begin
  let C₁ := B ∩ X,
  let C₂ := B \ X,
  have w : B.card = C₁.card + C₂.card, sorry,
  rw w at hB,
  rcases snorg hB with hC₁|hC₂,
  { refine ⟨C₁, sorry, hC₁, or.inl sorry⟩, },
  { refine ⟨C₂, sorry, hC₂, or.inr sorry⟩, },
end
```
which I think will give a nicer proof.

Still it's not really the right statement --- it's actually irrelevant here that `B X Y` are finsets! They could be any sets at all (we'd have to use `cardinal.mk` instead of `finset.card`) and its still true.

Reply:
Ideally we would start with the proof of
```
lemma foo {α : Type*} {n : ℕ} (h : cardinal.mk α ≥ 2 * n + 1) (s : set α) :
  cardinal.mk s ≥ n + 1 ∨ cardinal.mk (sᶜ : set α) ≥ n + 1 :=
sorry
```
and deduce this result from that one.